pkg/controller/clusterdeployment/clusterdeployment_controller.go

```diff
@@ -1659,3 +1675,69 @@ func (r *ReconcileClusterDeployment) deleteStaleProvisions(provs []*hivev1.Clust
 		}
 	}
 }
+
+// getAllSyncSetInstances return all syncset instances for a cluster deployment
+func (r *ReconcileClusterDeployment) getAllSyncSetInstances(cd *hivev1.ClusterDeployment) ([]*hivev1.SyncSetInstance, error) {
+
+	// get all syncset instances for the clusterdeployment namespace
+	ssiList := &hivev1.SyncSetInstanceList{}
+	err := r.Client.List(context.TODO(), ssiList, client.InNamespace(cd.Namespace))
+	if err != nil {
+		return nil, err
+	}
+
+	syncSetInstances := []*hivev1.SyncSetInstance{}
+	for i, syncSetInstance := range ssiList.Items {
+		if syncSetInstance.Spec.ClusterDeployment.Name == cd.Name {
+			syncSetInstances = append(syncSetInstances, &ssiList.Items[i])
+		}
+	}
+	return syncSetInstances, nil
+}
+
+// checkFailureSyncCondition return true when syncsetinstance status condition has hivev1.ApplyFailureSyncCondition
+func checkFailureSyncCondition(syncSetInstances []*hivev1.SyncSetInstance) bool {
+	for _, syncSetInstance := range syncSetInstances {
+		for _, condition := range syncSetInstance.Status.Conditions {
+			if condition.Type == hivev1.ApplyFailureSyncCondition {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// updateSyncSetFailureCondition returns true when it sets the hivev1.SyncSetFailureCondition
+// hivev1.SyncSetFailureCondition is set if any syncsetinstance has status condition type as hivev1.ApplyFailureSyncCondition will return true
+func (r *ReconcileClusterDeployment) updateSyncSetFailureCondition(cd *hivev1.ClusterDeployment, cdLog log.FieldLogger) (bool, error) {
+
+	// get all syncset instances for this cluster deployment
+	syncSetInstances, err := r.getAllSyncSetInstances(cd)
+	if err != nil {
+		cdLog.WithError(err).Error("unable to list related sync set instances for cluster deployment")
+		return false, err
+	}
+
+	isfailedCondition := checkFailureSyncCondition(syncSetInstances)
+
+	// if failure found, add the cluster deployment status condition SyncSetFailureCondition
+	if isfailedCondition == true {
+		cd.Status.Conditions = controllerutils.SetClusterDeploymentCondition(cd.Status.Conditions, hivev1.SyncSetFailureCondition,
+			corev1.ConditionTrue, "SyncSetApplyFailure", "One of the SyncSetInstance apply has failed", controllerutils.UpdateConditionAlways)
```

Comment:
will UpdateConditionAlways put us in a hot loop by always updating the LastProbeTime?

Reply:
also, since ths function (updateSyncSetFailureCondition) is returning the bool indicating whether the condition was modified, you might want to look into the SetClusterDeploymentConditionWithChangeCheck() call since it will handle notifying you if any updates were made.

Reply:
> will UpdateConditionAlways put us in a hot loop by always updating the LastProbeTime?

Right, I had plan to add the check existing value and update only if required.

Reply:
the SetClusterDeploymentConditionWithChangeCheck() will do a lot of the work of only updating when necessary and notifying you if something changed (so you know whether to call the Status().Update() ). is there any particular reason to replicate some of the functionality in that call here instead of just calling SetClusterDeploymentConditionWithChangeCheck()?